src/lib/core/trigger.cc

```diff
@@ -50,3 +73,65 @@ trigger_run_reverse(struct rlist *list, void *event)
 			return -1;
 	return 0;
 }
+
+void
+trigger_fiber_run(struct rlist *list, void *event, double timeout)
+{
+	struct trigger *trigger, *tmp;
+	unsigned trigger_count = 0;
+	struct region *region = &fiber()->gc;
+	RegionGuard guard(region);
+	bool expired = false;
+	struct ev_timer timer;
+	ev_timer_init(&timer, trigger_fiber_run_timeout, timeout, 0);
+	timer.data = &expired;
+	ev_timer_start(loop(), &timer);
+
+	/*
+	 * Calculating the total number of triggers.
+	 */
+	rlist_foreach_entry(trigger, list, link)
+		trigger_count++;
+
+	struct fiber **fibers = (struct fiber **)
+		region_alloc(region, trigger_count * sizeof(struct fiber *));
+	if (fibers == NULL) {
+		say_error("Failed to allocate %lu bytes in region_alloc "
+			  "for fiber", trigger_count * sizeof(struct fiber));
+		ev_timer_stop(loop(), &timer);
+		return;
+	}
+
+	unsigned current_fiber = 0;
+	rlist_foreach_entry_safe(trigger, list, link, tmp) {
+		char name[FIBER_NAME_INLINE];
+		snprintf(name, FIBER_NAME_INLINE,
+			 "trigger_fiber%d", current_fiber);
+		fibers[current_fiber] = fiber_new(name, trigger_fiber_f);
+		if (fibers[current_fiber] != NULL) {
+			fiber_set_joinable(fibers[current_fiber], true);
+			fiber_start(fibers[current_fiber], trigger, event);
+			current_fiber++;
+		} else {
+			diag_log();
+			diag_clear(diag_get());
+			ev_timer_stop(loop(), &timer);
+			return;
+		}
+	}
+
+	/*
+	 * Waiting for all triggers completion.
+	 */
+	for (unsigned int i = 0; i < current_fiber && ! expired; i++) {
+		if (fiber_join_timeout(fibers[i], timeout) != 0) {
+			assert(! diag_is_empty(diag_get()));
+			diag_log();
+			diag_clear(diag_get());
+		}
+	}
+	if (expired)
+		say_error("Time reserved for on_shutdown triggers work has "
```

Comment:
Braces are missing (in case if-body consists of several lines, we always wrap it into {}).
Also I'd edit message:
"on_shutdown triggers are timed out: not all triggers might have finished yet"